[wasm][tests] Enable line numbers in stack traces

radical opened the pull request:
Library tests don't show line numbers, essentially because the .pdb
files never become available to the `WasmAppBuilder` .

This is enabled by default for `Debug` builds only.

To use it with `Release` builds, add `/p:DebuggerSupport=true` to the
command line.

With the patch:

```
  fail: [FAIL] System.Reflection.Tests.MemberInfoTests.HasSameMetadataDefinitionAs__CornerCase_HasElementTypes
  info: Assert.All() Failure: 10 out of 10 items in the collection did not pass.
  info: [9]: Item: System.Double*
  info:      Xunit.Sdk.AllException: Assert.All() Failure: 2 out of 10 items in the collection did not pass.
  info:      [7]: Item: System.Double&
  info:           Xunit.Sdk.TrueException: Assert.True() Failure
  info:           Expected: True
  info:           Actual:   False
  info:              at Xunit.Assert.True(Nullable`1 condition, String userMessage) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\BooleanAsserts.cs:line 95
  info:              at Xunit.Assert.True(Boolean condition) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\BooleanAsserts.cs:line 62
  info:              at System.Reflection.Tests.MemberInfoTests.<>c__DisplayClass31_1.<HasSameMetadataDefinitionAs__CornerCase_HasElementTypes>b__1(Type t2) in /Users/radical/dev/runtime/src/libraries/System.Reflection/tests/MemberInfoTests.cs:line 491
  info:              at Xunit.Assert.All[Type](IEnumerable`1 collection, Action`1 action) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\CollectionAsserts.cs:line 36
```

akoeplinger commented on eng/testing/tests.wasm.targets:
```diff
@@ -79,4 +81,17 @@
+
+  <!-- linker automatically picks up the .pdb files, but they are not added to the publish list.
+       Add them explicitly here, so they can be used with WasmAppBuilder -->
+  <Target Name="AddPdbFilesToPublishList" AfterTargets="ILLink">
```
should we make this conditional on `DebuggerSupport == true`?

akoeplinger replied:
I looked at the sdk targets and now I'm confused because if I'm reading this correctly pdbs should already be added to ResolvedFileToPublish by default: https://github.com/dotnet/sdk/blob/40f28f2de8dac1b2297a73b072dbc08f27a181a6/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L558-L564

radekdoulik replied:
Maybe these pdb files are in different location after linking? I didn't find how the `_DebugSymbolsIntermediatePath` is set, so I am not sure.

radical replied:
> I looked at the sdk targets and now I'm confused because if I'm reading this correctly pdbs should already be added to ResolvedFileToPublish by default: https://github.com/dotnet/sdk/blob/40f28f2de8dac1b2297a73b072dbc08f27a181a6/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L558-L564

That's copying the pdb for the main assembly only. I'm adding them for the files from other locations, like copy local ones (in this case from the runtime pack).

akoeplinger replied:
Ah that makes sense, thanks.

akoeplinger commented on eng/testing/tests.wasm.targets:
```diff
@@ -27,7 +27,7 @@
     <EnableUnsafeUTF7Encoding>false</EnableUnsafeUTF7Encoding>
     <HttpActivityPropagationSupport>false</HttpActivityPropagationSupport>
-    <DebuggerSupport>false</DebuggerSupport>
+    <DebuggerSupport>true</DebuggerSupport>
```
I wonder if we should really change the default here since we want to match what Blazor sets by default, i.e. `UseDefaultBlazorWASMFeatureSwitches`.

akoeplinger replied:
actually, looks like Blazor defaults it to true when configuration is debug and provides an override:
https://github.com/dotnet/sdk/blob/1e4913f32344ce69ee17229168906140e8f88f0d/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets#L59

maybe we should do the same?

radical replied:
What do we use `UseDefaultBlazorWASMFeatureSwitches` for?

`<DebuggerSupport Condition="'$(DebuggerSupport)' == '' and '$(Configuration)' != 'Debug'">false</DebuggerSupport>`

Hm this would mean that a dev would need to explicitly set `/p:DebuggerSupport=true` to get the line numbers, when building in `Release`. I was hoping for line numbers in test runs too.

radical replied:
`eng/testing/tests.mobile.targets:    <UseDefaultBlazorWASMFeatureSwitches Condition="'$(UseDefaultBlazorWASMFeatureSwitches)' == ''">true</UseDefaultBlazorWASMFeatureSwitches>`

This defaults to true, and I don't see it being set in anywhere in our repo. So, is the intention to always test with this?
The question is whether it is OK to use `DebuggerSupport=true` by default then?

akoeplinger replied:
I don't know the history of `UseDefaultBlazorWASMFeatureSwitches` but presumably we wanted to actually test the same configuration that Blazor will use in the end.

I'd love to have line numbers by default but I don't know if that changes runtime behavior in a way that makes it risky, maybe ask Zoltan/Aleksey.

If we decide to do it I think we should move this property out of the `UseDefaultBlazorWASMFeatureSwitches` propertygroup since it won't be matching Blazor anymore and cause confusion.

radical committed:
[wasm] WasmAppBuilder: Copy pdb files whenever debugging

.. is enabled, which is for WasmDebugLevel != 0

radical committed:
[wasm][tests] Enable line numbers in stack traces

Library tests don't show line numbers, essentially because the .pdb
files never become available to the WasmAppBuilder .

This is enabled by default for `Debug` builds only.

To use it with `Release` builds, add `/p:DebuggerSupport=true` to the
command line.

With the patch:

```
  fail: [FAIL] System.Reflection.Tests.MemberInfoTests.HasSameMetadataDefinitionAs__CornerCase_HasElementTypes
  info: Assert.All() Failure: 10 out of 10 items in the collection did not pass.
  info: [9]: Item: System.Double*
  info:      Xunit.Sdk.AllException: Assert.All() Failure: 2 out of 10 items in the collection did not pass.
  info:      [7]: Item: System.Double&
  info:           Xunit.Sdk.TrueException: Assert.True() Failure
  info:           Expected: True
  info:           Actual:   False
  info:              at Xunit.Assert.True(Nullable`1 condition, String userMessage) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\BooleanAsserts.cs:line 95
  info:              at Xunit.Assert.True(Boolean condition) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\BooleanAsserts.cs:line 62
  info:              at System.Reflection.Tests.MemberInfoTests.<>c__DisplayClass31_1.<HasSameMetadataDefinitionAs__CornerCase_HasElementTypes>b__1(Type t2) in /Users/radical/dev/runtime/src/libraries/System.Reflection/tests/MemberInfoTests.cs:line 491
  info:              at Xunit.Assert.All[Type](IEnumerable`1 collection, Action`1 action) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\CollectionAsserts.cs:line 36
```

radical commented:
Last few builds seem to be just stuck!
